_posts/2017-11-22-macros.md

```diff
@@ -0,0 +1,305 @@
+untyped children.
+In scala-reflect, the burden is on the macro author to ensure this invariant
+holds.
+Mistakes from mixing untyped and typed tree nodes result in cryptic compiler
+crashes.
+In the new macro system, we want to avoid exposing such bad experience to macro
+authors and library users.
+
+This observation that tree transformation are difficult to support doesn't
+imply that new macros won't support them.
+It simply means that we need give this problem more time and thought so that we
+can come up with a robust solution.
+So far, we have seen two promising ideas to address or circumvent this problem,
+one by Martin Odersky and one by Lionel Parreaux.
+
+## Unification of LMS and Macros
+
+Martin Odersky recently shared [a gist](https://gist.github.com/odersky/f91362f6d9c58cc1db53f3f443311140)
+on Principled Metaprogramming for Scala.
+The design proposed in this gist is quite close to Meta OCaml as well as to a
+calculus by Davies and Pfenning inspired by modal logic, see
+["A modal analysis of staged computation"](https://dl.acm.org/citation.cfm?doid=382780.382785).
+
+This proposal has triggered a rich discussion in the gist, which you may find interesting.
+However, the proposal may go too far from current scala-reflect macros to
+cause a migration concern.
+Quoting Martin Odersky from
+[r/scala](https://www.reddit.com/r/scala/comments/7emo4s/oderskys_ideas_about_a_new_design_for_macros_that/dq6m1da/)
+
+> One attractive aspect of the system is that it works equally well for staging
+> and for macros.  But as a macro system it is quite restrictive because it
+> imposes a strong typing discipline, only works for full expressions, and in
+> its original version does not allow for code inspection. Lionel's Parreaux's
+> Squid system (paper to appear at upcoming POPL) is more powerful in that it
+> does allow inspection.
+
+Lionel Parreaux ([@LPTK]) is a PhD student at EPFL who has been working on
+related problems for the past couple years as part of his research on the
+intersection of database systems, programming languages and large-scala data
+analysis.
+
+## Squid: reusable and type-safe quasiquotes
+
+[Squid] is a metaprogramming framework that facilitates the type-safe
+manipulation of Scala programs through quasiquotes.
+Unlike scala-reflect quasiquotes, Squid quasiquotes are statically guaranteed
+to be well-typed, well-scoped and hygienic.
+The properties of Squid quasiquotes make them ideal for robust and safe
+transformation of trees, the exact problem we struggled with in our prototype
+macro system.
+
+One important attribute of Squid quasiquotes is that they are reusable.
+Similar to scala-reflect quasiquotes, it's possible to write generic utility
+functions with Squid quasiquotes and compose them into more complex tree
+structures.
+Many alternative statically typed quasiquote systems do not support this
+ability to combine small quasiquote fragments into larger programs.
```

Comment:
I would say that most quasiquote systems support that. I think this is what 'quasi' means in this context –– differentiating simple quotes from "quasi-quotes" that can contain holes to be filled in later.

The _reusability_ of Squid quasiquotes comes from the fact that you can "plug in" (finally-tagless style) different intermediate representations, that are then used to represent the program fragments. Quasiquotes simply provide a generic type-safe interface to do the manipulation. As you said, similar to scala-reflect but type-safe.

Reply:
I see, I guess I misunderstood the "reusable" aspect then. I had in mind reusability compared with `reify` where it's more tricky to write utility functions. Let me jst remove this whole paragraph since the point I was trying to make is kind of lost.